src/ucp/core/ucp_context.c

```diff
@@ -1768,15 +1721,89 @@ static ucs_status_t ucp_fill_aux_tls(ucs_string_set_t *aux_tls)
     return UCS_OK;
 }
 
+static void
+ucp_update_memtype_md_map(uint64_t mem_types_map, ucs_memory_type_t mem_type,
+                          ucp_md_index_t md_index, ucp_md_map_t *md_map_p)
+{
+    if (mem_types_map & UCS_BIT(mem_type)) {
+        *md_map_p |=  UCS_BIT(md_index);
+    }
+}
+
 static void ucp_fill_resources_reg_md_map_update(ucp_context_h context)
 {
     UCS_STRING_BUFFER_ONSTACK(strb, 256);
+    ucp_md_map_t reg_block_md_map    = 0;
+    ucp_md_map_t reg_nonblock_md_map = 0;
     ucs_memory_type_t mem_type;
     ucp_md_index_t md_index;
+    const uct_md_attr_v2_t *md_attr;
+
+    for (md_index = 0; md_index < context->num_mds; ++md_index) {
+        md_attr = &context->tl_mds[md_index].attr;
+        if (md_attr->flags & UCT_MD_FLAG_EXPORTED_MKEY) {
+            context->export_md_map |= UCS_BIT(md_index);
+        }
+
+        if (md_attr->flags & UCT_MD_FLAG_REG_DMABUF) {
+            context->dmabuf_reg_md_map |= UCS_BIT(md_index);
+        }
+    }
 
-    /* If we have a dmabuf provider for a memory type, it means we can register
-     * memory of this type with any md that supports dmabuf registration. */
     ucs_memory_type_for_each(mem_type) {
+        for (md_index = 0; md_index < context->num_mds; ++md_index) {
+            md_attr = &context->tl_mds[md_index].attr;
+            if (md_attr->dmabuf_mem_types & UCS_BIT(mem_type)) {
+                /* In case of multiple providers, take the first one */
+                if (context->dmabuf_mds[mem_type] == UCP_NULL_RESOURCE) {
+                    context->dmabuf_mds[mem_type] = md_index;
+                }
+            }
+
+            if (!(md_attr->flags & UCT_MD_FLAG_REG)) {
+                continue;
+            }
+
+            ucp_update_memtype_md_map(
+                    md_attr->reg_nonblock_mem_types, mem_type,
+                    md_index, &reg_nonblock_md_map);
```

Comment:
can we improve the logic here, so that instead of going over all mem types and then over all mds, we would go over all mds, and for each md we go over the relevant mem types for each thing - for example for each md iterate over all dmabuf_mem_types. instead of checking with UCS_BIT all possible types.